apt/doc/apt-config.8
.TH "apt-config" "8" "25 Oct 2000" "apt" "" 
.SH "NAME" 
apt-config \- APT Configuration Query program
.PP 
.SH "SYNOPSIS" 
apt-config command
.PP 
.SH "DESCRIPTION" 
\fBapt-config\fP is an internal program used by various portions of the APT
suite to provide consistent configurability\&. It accesses the main configuarion
file /etc/apt/apt\&.conf in a manner that is easy to use by scripted
applications\&.
.PP 
\fIcommand\fP is one of:
.IP o 
shell
.IP o 
dump
.PP 
Unless the -h, or --help option is given one of the above commands
must be present\&.
.PP 
.IP "\fBshell\fP" 
\fBshell\fP is used to access the configuration information from a shell script\&.
It is given pairs of arguments, the first being a shell variable and the
second the configuration value to query\&. As output it lists a series of shell
assignments commands for each present value\&. In a shell script it should be
used like:
.IP 

.nf 
 

OPTS="-f"

RES=`apt-config shell OPTS MyApp::Options`

eval $RES

.fi 
 

.IP 
This will set the shell environment variable $OPTS to the value of
MyApp::Options with a default of -f\&.
.IP 
If the configuration item to retrieve is prefixed with a / then it will
be retrieved using filename mode which prepends base paths\&.
.IP 
.IP "\fBdump\fP" 
Just show the contents of the configuration space\&.
.IP 
.PP 
.SH "OPTIONS" 
All command line options may be set using the configuration file, the
descriptions indicate the configuration option to set\&. For boolean
options you can override the config file by using something like \fB-f-\fP,
\fB--no-f\fP, \fB-f=no\fP or several other variations\&.
.PP 
.IP "\fB-h, --help\fP" 
Show a short usage summary\&.
.IP 
.IP "\fB-v, --version\fP" 
Show the program verison\&.
.IP 
.IP "\fB-c, --config-file\fP" 
Configuration File; Specify a configuration file to use\&. \fBapt-get\fP will
read the default configuration file and then this configuration file\&. See
\fBapt\&.conf(5)\fP for syntax information\&.
.IP 
.IP "\fB-o, --option\fP" 
Set a Configuration Option; This will set an arbitary configuration option\&.
The syntax is 

.nf 
 
-o Foo::Bar=bar
.fi 
 

.PP 
.SH "SEE ALSO" 
apt\&.conf(5)
.PP 
.SH "DIAGNOSTICS" 
apt-config returns zero on normal operation, decimal 100 on error\&.
.PP 
.SH "BUGS" 
See http://bugs\&.debian\&.org/apt\&.  If you wish to report a
bug in \fBapt-config\fP, please see \fB/usr/doc/debian/bug-reporting\&.txt\fP
or the \fBbug(1)\fP command\&. If you are using apt on a RPM based
system, please use http://distro\&.conectiva\&.com\&.br/bugzilla/\&.
.PP 
.SH "AUTHOR" 
apt-get was written by the APT team <apt@packages\&.debian\&.org>
and ported to RPM based systems by Conectiva S.A. 
<kojima@conectiva\&.com\&.br>\&.


apt/cmdline/apt-config.cc
// -*- mode: cpp; mode: fold -*-
// Description								/*{{{*/
// $Id: apt-config.cc,v 1.3 2001/08/01 21:57:05 kojima Exp $
/* ######################################################################
   
   APT Config - Program to manipulate APT configuration files
   
   This program will parse a config file and then do something with it.
   
   Commands:
     shell - Shell mode. After this a series of word pairs should occure.
             The first is the environment var to set and the second is
             the key to set it from. Use like: 
 eval `apt-config shell QMode apt::QMode`
   
   ##################################################################### */
									/*}}}*/
// Include Files							/*{{{*/
#include <apt-pkg/cmndline.h>
#include <apt-pkg/error.h>
#include <apt-pkg/init.h>
#include "config.h"

#include <i18n.h>

#include <iostream>
									/*}}}*/

// DoShell - Handle the shell command					/*{{{*/
// ---------------------------------------------------------------------
/* */
bool DoShell(CommandLine &CmdL)
{
   for (const char **I = CmdL.FileList + 1; *I != 0; I += 2)
   {
      if (I[1] == 0 || strlen(I[1]) == 0)
	 return _error->Error(_("Arguments not in pairs"));

      // Check if the caller has requested a directory path
      if (I[1][strlen(I[1])-1] == '/')
      {
	 char S[300];
	 strcpy(S,I[1]);
	 S[strlen(S)-1] = 0;
	 if (_config->Exists(S) == true)
	    cout << *I << "=\"" << _config->FindDir(S) << '"' << endl;
      }
      
      if (_config->Exists(I[1]) == true)
	 cout << *I << "=\"" << _config->Find(I[1]) << '"' << endl;
   }
   
   return true;
}
									/*}}}*/
// DoDump - Dump the configuration space				/*{{{*/
// ---------------------------------------------------------------------
/* */
bool DoDump(CommandLine &CmdL)
{
   _config->Dump();
   return true;
}
									/*}}}*/
// ShowHelp - Show the help screen					/*{{{*/
// ---------------------------------------------------------------------
/* */
int ShowHelp()
{
   cout << PACKAGE << ' ' << VERSION << " for " << COMMON_CPU <<
       " compiled on " << __DATE__ << "  " << __TIME__ << endl;
   if (_config->FindB("version") == true)
      return 100;
   
   cout << _("Usage: apt-config [options] command") << endl;
   cout << endl;
   cout << _("apt-config is a simple tool to read the APT config file") << endl;   
   cout << endl;
   cout << _("Commands:") << endl;
   cout << _("   shell - Shell mode") << endl;
   cout << _("   dump - Show the configuration") << endl;
   cout << endl;
   cout << _("Options:") << endl;
   cout << _("  -h   This help text.") << endl;
   cout << _("  -c=? Read this configuration file") << endl;
   cout << _("  -o=? Set an arbitary configuration option, eg -o dir::cache=/tmp") << endl;
   return 100;
}
									/*}}}*/

int main(int argc,const char *argv[])
{
   CommandLine::Args Args[] = {
      {'h',"help","help",0},
      {'v',"version","version",0},
      {'c',"config-file",0,CommandLine::ConfigFile},
      {'o',"option",0,CommandLine::ArbItem},
      {0,0,0,0}};
   CommandLine::Dispatch Cmds[] = {{"shell",&DoShell},
                                   {"dump",&DoDump},
                                   {0,0}};
    
   setlocale(LC_ALL, "");
   bindtextdomain(PACKAGE, LOCALEDIR);
   textdomain(PACKAGE);
   
   // Parse the command line and initialize the package library
   CommandLine CmdL(Args,_config);
   if (pkgInitialize(*_config) == false ||
       CmdL.Parse(argc,argv) == false)
   {
      _error->DumpErrors();
      return 100;
   }

   // See if the help should be shown
   if (_config->FindB("help") == true ||
       CmdL.FileSize() == 0)
      return ShowHelp();

   // Match the operation
   CmdL.DispatchArg(Cmds);
   
   // Print any errors or warnings found during parsing
   if (_error->empty() == false)
   {
      bool Errors = _error->PendingError();
      _error->DumpErrors();
      return Errors == true?100:0;
   }
   
   return 0;
}
